test/Concurrency/isolated_default_arguments.swift

```diff
@@ -0,0 +1,182 @@
+  // expected-error@+1 {{main actor-isolated default argument cannot be synchronously evaluated from a nonisolated context}}
+  mainActorDefaultArg()
+
+  // expected-error@+1 {{main actor-isolated default argument cannot be synchronously evaluated from a nonisolated context}}
+  mainActorClosure()
+
+  // expected-error@+1 {{main actor-isolated default argument cannot be synchronously evaluated from a nonisolated context}}
+  mainActorClosureCall()
+
+  await mainActorDefaultArg(value: requiresMainActor())
+
+  // 'await' doesn't help closures in default arguments; the calling context needs a matching
+  // actor isolation for that isolation to be inferred for the closure value.
+
+  // expected-error@+2 {{main actor-isolated default argument cannot be synchronously evaluated from a nonisolated context}}
+  // expected-warning@+1 {{no 'async' operations occur within 'await' expression}}
+  await mainActorClosure()
+
+  // expected-error@+2 {{main actor-isolated default argument cannot be synchronously evaluated from a nonisolated context}}
+  // expected-warning@+1 {{no 'async' operations occur within 'await' expression}}
+  await mainActorClosureCall()
+}
+
+func conflictingIsolationDefaultArg(
+  // expected-error@+1 {{default argument cannot be both main actor-isolated and global actor 'SomeGlobalActor'-isolated}}
+  value: (Int, Int) = (requiresMainActor(), requiresSomeGlobalActor())
+) {}
+
+func closureLosesIsolation(
+  closure: @Sendable () -> Void = {
+    // expected-note@+1 {{calls to local function 'f()' from outside of its actor context are implicitly asynchronous}}
+    @MainActor func f() {}
+    // expected-error@+1 {{call to main actor-isolated local function 'f()' in a synchronous nonisolated context}}
+    f()
+  }
+) {}
+
+func closureIsolationMismatch(
+  closure: @SomeGlobalActor () -> Void = {
+    // expected-note@+1 {{calls to local function 'f()' from outside of its actor context are implicitly asynchronous}}
+    @MainActor func f() {}
+    // expected-error@+1 {{call to main actor-isolated local function 'f()' in a synchronous global actor 'SomeGlobalActor'-isolated context}}
+    f()
+  }
+) {}
+
+func conflictingClosureIsolation(
+  // expected-error@+1 {{default argument cannot be both main actor-isolated and global actor 'SomeGlobalActor'-isolated}}
+  closure: () -> Void = {
+    _ = requiresMainActor()
+    _ = requiresSomeGlobalActor()
+  }
+) {}
+
+func isolationInLocalFunction(
+  closure: () -> Void = {
+    nonisolated func local() -> Int {
+      // expected-error@+1 {{call to main actor-isolated global function 'requiresMainActor()' in a synchronous nonisolated context}}
+      requiresMainActor()
+    }
+  }
+) {}
+
+actor A {}
+
+func closureWithIsolatedParam(
+  closure: (isolated A) -> Void = { _ in
+    // expected-error@+1 {{call to main actor-isolated global function 'requiresMainActor()' in a synchronous actor-isolated context}}
+    _ = requiresMainActor()
+  }
+) {}
```

Comment:
I guess it is covered implicitly, but maybe worth adding test for 

```
extension A { 
func closureWithIsolatedParam(
   closure: () -> Void = { _ in
     _ = requiresMainActor()
   }
 ) {}
}
```

that's somewhat interesting... that would fail when called inside actor, but across could actually work...? Hop to main for param, then hop to invoke method...? I'm not sure what would happen today tbh, so may be worth the test-case

Reply:
Default arguments are always evaluated in the caller's isolation domain, so using a MainActor-isolated default argument to an actor-isolated method is conceptually "fine" - it only has implications on where you can use the default argument from. However, I'm considering an (artificial) restriction that the required isolation has to be either `nonisolated` or the same isolation as the function itself because I'm not sure why you would want a default argument that has any other isolation.

Reply:
Gotcha that makes sense; I'm not sure about the artificial restriction, I guess things fall into place and can work without it 🤔